### PR TITLE
WebSocket session management

### PR DESCRIPTION
**Pull request by teivah**
I have not been able to reproduce the problem. Yet, I noticed that the management of the `session` object in `WebSocket` was not correctly synchronized throughout the whole chain. Moreover, under high-throughput I managed to produce the following exception:

```java
Jan 29, 2019 3:34:32 PM org.glassfish.grizzly.nio.transport.TCPNIOTransport$DefaultChannelConfigurator postConfigure
WARNING: GRIZZLY0004: Can not set TCP_NODELAY to true
java.net.SocketException: Invalid argument
``` 

The new implementation relies on Tyrus to handle reconnections and changes the way to synchronize the messages send.

### CODE REVIEW

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java [hunk 1]
```diff
@@ -105,10 +98,19 @@ Session connect(URI uri) {
...
+      container.getProperties().put(
+          ClientProperties.RECONNECT_HANDLER, InfiniteReconnectHandler.INSTANCE);
+
+      new SessionBuilder(container)
           .clientEndpointConfig(config)
           .uri(uri)
-          .onOpen((s, ec) -> log.info("Connected session {}", s.getId()))
+          .onOpen((s, ec) -> {
+            synchronized (this) {
+              System.out.println(s);
```
Comment (nahguam) [inline client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: remove sysout

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java [hunk 2]
```diff
@@ -135,13 +137,10 @@ Session connect(URI uri) {
...
   void send(Event event) {
-    connect();
     log.debug("Sending Event: {}", event);
-    synchronized (session) {
-      try {
-        session
-            .getBasicRemote()
-            .sendBinary(ByteBuffer.wrap(mapper.writeValueAsBytes(event)));
+    synchronized (this) {
+      try (OutputStream out = session.getBasicRemote().getSendStream()) {
```
Comment (nahguam) [inline client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: Does this revert the changes from your previous PR?

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java [hunk 3]
```diff
@@ -105,10 +98,19 @@ Session connect(URI uri) {
         container.getProperties().put(SSL_ENGINE_CONFIGURATOR, configurator);
       }
 
-      return new SessionBuilder(container)
+      container.getProperties().put(
+          ClientProperties.RECONNECT_HANDLER, InfiniteReconnectHandler.INSTANCE);
+
+      new SessionBuilder(container)
```
Comment (nahguam) [inline client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: should this be assigned to the `session`?

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java (hunk 1 above)
Comment (teivah) [reply client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: f*** sorry...

**Comment by nahguam** [comment]
please add a changelog entry

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java (hunk 2 above)
Comment (teivah) [reply client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: It was not merged yet

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java (hunk 3 above)
Comment (teivah) [reply client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: It's done here: https://github.com/HotelsDotCom/data-highway/blob/ef432cfd9b471c7c97453da28be59e56e6c513ab/client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java#L107

**Comment by teivah** [comment]
I might have missed something but I did not see the point to synchronize the `send()` method. In Tyrus, there's already something for this: https://github.com/tyrus-project/tyrus/blob/master/core/src/main/java/org/glassfish/tyrus/core/ProtocolHandler.java#L114 (used for example while sending a message: https://github.com/tyrus-project/tyrus/blob/master/core/src/main/java/org/glassfish/tyrus/core/ProtocolHandler.java#L278)

**Comment by teivah** [comment]
@nahguam Remarks handled

**Comment by teivah** [comment]
@nahguam Done (`synchronized` is back).

File: client/offramp/pom.xml [hunk 4]
```diff
@@ -14,6 +13,7 @@
 
   <properties>
     <tyrus.version>1.12</tyrus.version>
+    <spring-test.version>5.1.4.RELEASE</spring-test.version>
```
Comment (nahguam) [inline client/offramp/pom.xml]: You shouldn't need to declare the version as we import the Spring Boot dependencies https://github.com/HotelsDotCom/data-highway/blob/master/pom.xml#L151

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java [hunk 5]
```diff
@@ -216,6 +214,7 @@ Disposable startKeepAlives() {
   }
 
   interface Factory {
+
```
Comment (Noddy76) [inline client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: I am not a fan of these leading blank lines in class/interface definitions.

File: client/offramp/pom.xml (hunk 4 above)
Comment (teivah) [reply client/offramp/pom.xml]: Fixed.

File: client/offramp/src/main/java/com/hotels/road/offramp/client/OfframpClientImpl.java [hunk 6]
```diff
@@ -41,14 +46,16 @@
...
   private final @Getter(PACKAGE) MessageProcessor<T> messages;
   private final CommitHandler commitHandler;
-  private final @Getter(PACKAGE) DirectProcessor<Rebalance> rebalances = create();
+  private final @Getter(PACKAGE) DirectProcessor<Rebalance> rebalances = DirectProcessor.create();
   private final OfframpOptions<T> options;
   private final WebSocket socket;
 
   OfframpClientImpl(OfframpOptions<T> options) {
-    this(options, WebSocket::new, CommitHandler::new);
+    this(options,
```
[the remark this answers is not in the document]
Comment (teivah) [reply client/offramp/src/main/java/com/hotels/road/offramp/client/OfframpClientImpl.java]: Fixed.

File: client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java (hunk 5 above)
Comment (teivah) [reply client/offramp/src/main/java/com/hotels/road/offramp/client/WebSocket.java]: I can remove it manually but actually, it's part of this Hcom IntelliJ code style that I imported which adds this new line automatically.
@nahguam Did you experience the same thing?
